src/openms/source/ANALYSIS/ID/AhoCorasickDA.cpp

```diff
@@ -0,0 +1,1154 @@
+    // which is never a hit, or is still in state because a hit was found and otherwise always the same hit is
+    // reported.
+    // makes only the next valid transition, no test for hit
+    if (node_pos_ != 0)
+    {
+      if (isAmbiguous_(*protein_))
+      {
+        if (!consumeAmbiguousAA_(*protein_))
+        {
+          return false;
+        }
+      }
+      else
+      {
+        if (getNextNode_(node_pos_, code_(*protein_), node_pos_, max_depth_decrease_))
+        {
+          if (*protein_ != '\0')
+          {
+            ++prot_pos_;
+            ++protein_;
+          }
+        }
+        else
+        {
+          if (!useSpawn_())
+          {
+            return false;
+          }
+        }
+      }
+    }
+    while (retrieval_())
+    {
+      // Node itself is a hit
+      if (output_[node_pos_] != SUBSTR_IS_HIT)
+      {
+        peptide_index = output_[node_pos_];
+        pos_in_protein = prot_pos_ - sequences_[peptide_index].size();
+        multipleOccurrence_(peptide_index);
+        findSubstrOfHit_();
+      }
+        // Node itself is not a hit but a supply link leads to a hit
+      else
+      {
+        findSubstrOfHit_();
+        // Returns the last hit, all other hits are returned when calling findNext again
+        if (!multi_hit_.empty())
+        {
+          peptide_index = multi_hit_.back();
+          pos_in_protein = prot_pos_ - sequences_[peptide_index].size();
+          multi_hit_.pop_back();
+        }
+          // In the case of ambiguous aa, the supply link leading to a hit may drop the first ambiguous aa and nothing is returned
+        else
+        {
+          return findNext(pos_in_protein, peptide_index);
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+  //-------------------------------------------------------------------------------------------------------------------------------------
+  //  CDA - retrieval
+  //-------------------------------------------------------------------------------------------------------------------------------------
+  void AhoCorasickDA::findSubstrOfHit_()
+  {
+    Int32 sl_node = 0;
+    // Global variable should not be changed
+    uint16_t amb_depth = max_depth_decrease_;
+    SupplyLink_* node = getSupplyLink_(node_pos_);
+    // Follows the supply links to root
+    while ((node->link != 0) && followSupplyLink_(node, sl_node, amb_depth))
+    {
+      if (isHit_(sl_node))
+      {
+        // Test if node is an output node or just marked that a substr is a hit
+        if (output_[sl_node] != SUBSTR_IS_HIT)
+        {
+          multi_hit_.push_back(output_[sl_node]);
+          multipleOccurrence_(output_[sl_node]);
+        }
+      }
+      node = getSupplyLink_(sl_node);
+    }
+  }
+  void AhoCorasickDA::multipleOccurrence_(const Size pep_idx )
+  {
+    UInt32 idx = multi_idx_[pep_idx].first;
+    for (UInt32 i = 1; i <= multi_idx_[pep_idx].second; ++i )
```

Comment:
`multi_idx_[pep_idx].second` searches the datastructure multiple times (costly!)
better remember the value and store it once.